File: sdk-ts/scripts/publish.js
#!/usr/bin/env node

/**
 * Publish ArkRelay TypeScript SDK to NPM
 *
 * Usage:
 *   node scripts/publish.js --test    # Dry run
 *   node scripts/publish.js           # Actual publish
 *   node scripts/publish.js --version 0.2.0  # Bump version
 */

import { execSync } from 'child_process';
import { readFileSync, writeFileSync } from 'fs';
import { join } from 'path';
import { createRequire } from 'module';

const require = createRequire(import.meta.url);

function runCommand(cmd, description) {
  console.log(`\n🔄 ${description}`);
  console.log(`Running: ${cmd}`);

  try {
    const output = execSync(cmd, { encoding: 'utf8', stdio: 'pipe' });
    console.log(`✅ ${description} completed successfully`);
    if (output) {
      console.log(`Output: ${output}`);
    }
    return output;
  } catch (error) {
    console.error(`❌ Error: ${error.message}`);
    console.error(`❌ Command failed: ${cmd}`);
    process.exit(1);
  }
}

function bumpVersion(newVersion) {
  const packageJsonPath = join(process.cwd(), 'package.json');
  const packageJson = JSON.parse(readFileSync(packageJsonPath, 'utf8'));

  packageJson.version = newVersion;

  writeFileSync(packageJsonPath, JSON.stringify(packageJson, null, 2) + '\n');
  console.log(`📦 Version bumped to ${newVersion}`);
}

function main() {
  const args = process.argv.slice(2);
  const testMode = args.includes('--test');
  const versionArg = args.find(arg => arg.startsWith('--version'));

  // Bump version if specified
  if (versionArg) {
    const newVersion = versionArg.split('=')[1] || args[args.indexOf('--version') + 1];
    if (newVersion) {
      bumpVersion(newVersion);
    }
  }

  // Validate we're in the right directory
  if (!require('fs').existsSync('package.json')) {
    console.error('❌ Error: package.json not found. Run from sdk-ts directory.');
    process.exit(1);
  }

  const packageJson = JSON.parse(readFileSync('package.json', 'utf8'));
  const version = packageJson.version;

  console.log(`🚀 Publishing ArkRelay TypeScript SDK v${version}`);

  if (testMode) {
    console.log('🧪 Test mode - no actual publish');
  }

  // Clean previous builds
  runCommand('npm run clean', 'Cleaning previous builds');

  // Install dependencies
  runCommand('npm install', 'Installing dependencies');

  // Build the package
  runCommand('npm run build', 'Building package');

  // Run dry run to check
  runCommand('npm run publish:dry', 'Running dry run');

  // Check build output
  const distFiles = require('fs').readdirSync('dist');
  console.log(`📦 Built files: ${distFiles.join(', ')}`);

  // Publish to appropriate registry
  if (testMode) {
    console.log('\n🧪 Test publish skipped (use --publish to actually publish)');
    console.log('✅ Package is ready for publishing!');
  } else {
    runCommand('npm publish --access public', 'Publishing to NPM');
    console.log('\n🎉 Published to NPM!');
    console.log('Install with: npm install arkrelay-sdk');
  }

  // Show package contents
  console.log('\n📋 Package contents:');
  runCommand('npm pack --json', 'Showing package contents');
}

main();

File: sdk-py/scripts/publish.py
#!/usr/bin/env python3
"""
Publish ArkRelay Python SDK to PyPI

Usage:
    python scripts/publish.py --test    # Publish to Test PyPI
    python scripts/publish.py           # Publish to PyPI
"""

import argparse
import subprocess
import sys
from pathlib import Path


def run_command(cmd, description):
    """Run a command and handle errors"""
    print(f"\n🔄 {description}")
    print(f"Running: {cmd}")

    result = subprocess.run(cmd, shell=True, capture_output=True, text=True)

    if result.returncode != 0:
        print(f"❌ Error: {result.stderr}")
        print(f"❌ Command failed: {cmd}")
        sys.exit(1)
    else:
        print(f"✅ {description} completed successfully")
        if result.stdout:
            print(f"Output: {result.stdout}")


def main():
    parser = argparse.ArgumentParser(description="Publish ArkRelay Python SDK")
    parser.add_argument("--test", action="store_true", help="Publish to Test PyPI")
    parser.add_argument("--version", help="Version to publish (overrides current version)")
    args = parser.parse_args()

    # Validate we're in the right directory
    if not Path("pyproject.toml").exists():
        print("❌ Error: pyproject.toml not found. Run from sdk-py directory.")
        sys.exit(1)

    # Update version if specified
    if args.version:
        run_command(
            f"sed -i.bak 's/version = .*/version = \"{args.version}\"/' pyproject.toml",
            f"Updating version to {args.version}"
        )
        run_command("rm -f pyproject.toml.bak", "Cleaning up backup")

    # Clean previous builds
    run_command("rm -rf dist/ build/ *.egg-info", "Cleaning previous builds")

    # Install build dependencies
    run_command(
        "python -m pip install --upgrade build twine",
        "Installing build dependencies"
    )

    # Build the package
    run_command("python -m build", "Building package")

    # Check the package
    run_command("twine check dist/*", "Checking package")

    # Publish to appropriate repository
    if args.test:
        run_command(
            "twine upload --repository testpypi dist/*",
            "Publishing to Test PyPI"
        )
        print("\n🎉 Published to Test PyPI!")
        print("Install with: pip install --index-url https://test.pypi.org/simple/ arkrelay")
    else:
        run_command(
            "twine upload dist/*",
            "Publishing to PyPI"
        )
        print("\n🎉 Published to PyPI!")
        print("Install with: pip install arkrelay")


if __name__ == "__main__":
    main()